samples/extensions/khr/externalmemory/main.c

```diff
@@ -0,0 +1,692 @@
+    const char* const* const required_khronos_extensions,
+    const size_t required_khronos_extensions_count)
+{
+    size_t supported_extensions_count;
+    OCLERROR_PAR(clGetDeviceInfo(cl_device, CL_DEVICE_EXTENSIONS, 0, NULL,
+                                 &supported_extensions_count),
+                 NULL, err);
+    char* supported_extensions =
+        (char*)malloc(supported_extensions_count * sizeof(char));
+    OCLERROR_PAR(clGetDeviceInfo(cl_device, CL_DEVICE_EXTENSIONS,
+                                 supported_extensions_count,
+                                 supported_extensions, NULL),
+                 NULL, err);
+
+    for (size_t i = 0; i < required_khronos_extensions_count; ++i)
+    {
+        size_t j = 0;
+        while (strcmp(required_khronos_extensions[i], &supported_extensions[j])
+               && j < supported_extensions_count)
+        {
+            ++j;
+        }
+        if (j == supported_extensions_count)
+        {
+            return false;
+        }
+    }
+    return true;
+err:
+    return false;
+}
+
+// Check if a given OpenCL device supports a particular external memory handle
+// type.
+bool check_external_memory_handle_type(
+    const cl_device_id cl_device,
+    cl_external_memory_handle_type_khr cl_external_memory_handle_type)
+{
+    cl_external_memory_handle_type_khr* supported_handle_types;
+    size_t supported_handle_types_count = 0;
+
+    OCLERROR_PAR(
+        clGetDeviceInfo(cl_device,
+                        CL_DEVICE_EXTERNAL_MEMORY_IMPORT_HANDLE_TYPES_KHR, 0,
+                        NULL, &supported_handle_types_count),
+        NULL, err);
+    supported_handle_types = (cl_external_memory_handle_type_khr*)malloc(
+        supported_handle_types_count);
+
+    OCLERROR_PAR(
+        clGetDeviceInfo(
+            cl_device, CL_DEVICE_EXTERNAL_MEMORY_IMPORT_HANDLE_TYPES_KHR,
+            supported_handle_types_count, supported_handle_types, NULL),
+        NULL, err);
+    for (size_t i = 0; i < supported_handle_types_count; ++i)
+    {
+        if (cl_external_memory_handle_type == supported_handle_types[i])
+        {
+            return true;
+        }
+    }
+    return false;
+err:
+    fprintf(stderr,
+            "Error: OpenCL could not query supported external memory handle "
+            "types\n");
+    exit(EXIT_FAILURE);
+}
+
+int main(int argc, char* argv[])
+{
+    cl_int error = CL_SUCCESS;
+    cl_int end_error = CL_SUCCESS;
+    cl_platform_id cl_platform;
+    cl_device_id cl_device;
+    VkPhysicalDevice vk_physical_device;
+    VkDevice vk_device;
+    cl_context context = NULL;
+    cl_command_queue queue = NULL;
+
+    cl_program program;
+
+    // Parse command-line options.
+    struct cl_sdk_options_Diagnostic diag_opts = { .quiet = false,
+                                                   .verbose = false };
+    struct cl_sdk_options_SingleDevice dev_opts = {
+        .triplet = { 0, 0, CL_DEVICE_TYPE_ALL }
+    };
```

Comment:
I think it would be better not to include these options in this example, as it is already hard to find an OpenCL device compatible with Vulkan. These options add extra logic that I don't really see useful for this sample.